arbor/morph/em_morphology.cpp

```diff
@@ -0,0 +1,150 @@
+    }
+
+    // Cache the sample locations.
+    // Iterate backwards over branches distal to root, so that the parent branch at
+    // fork points will label its distal sample.
+    sample_locs_.resize(ns);
+    for (int b=nb-1; b>=0; --b) {
+        auto idx = util::make_range(morph_.branch_indexes(b));
+        double len = branch_lengths_[b];
+        // Handle 0 length branch.
+        len = len==0.? 1.: len;
+        double start = dist2root_[idx.front()];
+        for (auto i: idx) {
+            sample_locs_[i] = {msize_t(b), (dist2root_[i]-start)/len};
+        }
+        // For ensure that all non-spherical branches have their last sample 
+        if (idx.size()>1u) {
+            sample_locs_[idx.back()] = mlocation{msize_t(b), 1};
+        }
+    }
+    sample_locs_[0] = mlocation{0, 0.};
+
+    // Cache the location of terminal and fork points.
+    auto& props = morph_.sample_props();
+    for (auto i: count_along(props)) {
+        auto p = props[i];
+        if (is_terminal(p)) {
+            terminals_.push_back(sample2loc(i));
+        }
+        if (is_fork(p)) {
+            forks_.push_back(sample2loc(i));
+        }
+    }
+}
+
+const morphology& em_morphology::morph() const {
+    return morph_;
+}
+
+mlocation em_morphology::root() const {
+    return {0,0};
+}
+
+mlocation em_morphology::sample2loc(msize_t sid) const {
+    if (sid>=morph_.num_samples()) {
+        throw morphology_error(util::pprintf("Sample {} does not exist in morpology", sid));
+    }
+    return sample_locs_[sid];
+}
+
+mlocation_list em_morphology::terminals() const {
+    return terminals_;
+}
+
+mlocation_list em_morphology::cover(mlocation loc, bool include_loc) const {
+    mlocation_list L{};
+    if (include_loc) L.push_back(loc);
+
+    // If the location is not at the end of a branch, it is its own cover.
+    if (loc.pos>0. && loc.pos<1.) return L;
+
+    // The root is a special case.
+    if (loc==mlocation{0,0}) {
+        // Spherical root: nothing more to do
+        if (morph_.spherical_root()) return L;
+        // Non-spherical root: include start of each branch attached to the root
+        for (auto b: morph_.branch_children(mnpos)) {
+            if (b) L.push_back({b, 0});
+        }
+        return L;
+    }
+
+    if (loc.pos==1) {
+        // The location is at the end of a branch: at the location at
+        // the start of each child branch.
+        for (auto b: morph_.branch_children(loc.branch)) {
+            L.push_back({b, 0});
+        }
+    }
+    else if (loc.pos==0) {
+        // The location is at the start of a branch: at the location at
+        // the end of the parent branch, and locations at the start of
+        // each child of the parent branch.
+        auto p = morph_.branch_parent(loc.branch);
+        if (p!=mnpos) L.push_back({p, 1});
+        for (auto b: morph_.branch_children(p)) {
+            if (b!=loc.branch) L.push_back({b, 0});
+        }
+    }
```

Comment:
This case and the `loc==mlocation{0,0}` case for non-spherical root could be combined, which would avoid a bit of code duplication (and of course make things a little tighter.). You could e.g. change the root special case code to just
```
if (loc==mlocation{0,0} && morph_.spherical_root()) return L;
```

Reply:
Done.